### ISSUE TITLE
macOS: OpenCV 3.4.1 exception on simple example with cvSaveImage()

### ISSUE THREAD

**Issue by maelvalais** [opencv/opencv#11076]
##### System information (version)

- OpenCV => 3.4.1 (I am using Homebrew: the version is [3.4.1_2])
- Operating System / Platform => x86_64-apple-darwin17.4.0 (macOS 10.13.3)
- Compiler => Apple LLVM version 9.0.0 (clang-900.0.39.2)

##### Detailed description

Using `cvSaveImage()` on macOS (using Homebrew's OpenCV 3.4.1) produces an exception in `matrix_wrap.cpp`.

##### Steps to reproduce

Save the following snippet into `main.cc`:
```.cpp
#include "opencv2/core/core_c.h"
#include "opencv2/imgcodecs/imgcodecs_c.h"

int main(int argc, char **argv)
{
    printf("loading: %s\n", argv[1]);
    IplImage *img = cvLoadImage(argv[1]);

    printf("saving: %s\n", argv[2]);
    cvSaveImage(argv[2], img);
}
```

compiled with:

    gcc main.cc $(pkg-config --cflags --libs opencv)

download the following image and name it `test1.png`:

![test1](https://user-images.githubusercontent.com/2195781/37419433-c31f63ee-27b4-11e8-8e10-b8a57fd9cad2.png)

Then run

    ./a.out test1.png out.jpg

The following error happens:

```
loading: test.png
saving: out.jpg
OpenCV(3.4.1) Error: Assertion failed ((flags & FIXED_TYPE) != 0) in type, file /tmp/opencv-20180307-60086-ryy1b3/opencv-3.4.1/modules/core/src/matrix_wrap.cpp, line 807
libc++abi.dylib: terminating with uncaught exception of type cv::Exception: OpenCV(3.4.1) /tmp/opencv-20180307-60086-ryy1b3/opencv-3.4.1/modules/core/src/matrix_wrap.cpp:807: error: (-215) (flags & FIXED_TYPE) != 0 in function type

[1]    3082 abort      ./a.out test.png out.jpg
```

Using OpenCV 3.4.0_1 (from Homebrew, again) works as expected.

Anyone having the same issue? Or is it only me?

[3.4.1_2]: https://github.com/Homebrew/homebrew-core/commits/971c417e813dc71f7a8e3eef5122d932dcab48e1/Formula/opencv.rb

**Comment by alalek** [comment]
OpenCV C API is deprecated and broken. You should not use OpenCV with C compiler, even with OpenCV 1.0 API.
Refer to OpenCV [samples](https://github.com/opencv/opencv/tree/3.4.1/samples/cpp) for code reference.

**Comment by maelvalais** [comment]
Thank you for the answer!! 🙂

So I changed to
```.cpp
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/core/utility.hpp>
using namespace cv;

int main(int argc, char **argv)
{
    printf("loading: %s\n", argv[1]);
    Mat img = imread(argv[1],1);

    printf("saving: %s\n", argv[2]);
    imwrite(argv[2], img);
}
```

And compiling with

    clang++ main.cc $(pkg-config --cflags --libs opencv)

it works perfectly. Thanks!

**Comment by jtlz2** [comment]
I'm getting this too @hawkWei07 -

$ ipython
Python 2.7.15 |Anaconda custom (64-bit)| (default, May  1 2018, 18:37:05)
Type "copyright", "credits" or "license" for more information.

IPython 5.8.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import cv2

In [2]: cv2.__version__
Out[2]: '3.4.3'

In [3]: cv2.ORB_create().save('test.yaml')
libc++abi.dylib: terminating with uncaught exception of type cv::Exception: OpenCV(3.4.3) /Users/travis/build/skvark/opencv-python/opencv/modules/core/src/persistence_yml.cpp:785: error: (-2:Unspecified error) EndWriteStruct w/o matching StartWriteStruct in function 'icvYMLEndWriteStruct'

Abort trap: 6

**Comment by alalek** [comment]
`cvSaveImage()` is fixed here: #19203 (3.4 branch only)
